common/persistence/visibility/store/elasticsearch/processor_test.go

```diff
@@ -219,6 +228,40 @@ func (s *processorSuite) TestAdd_ConcurrentAdd_Duplicates() {
 	s.Equal(1, s.esProcessor.mapToAckFuture.Len(), "only one request should be in the bulk")
 }
 
+func (s *processorSuite) TestAdd_ConcurrentAdd_Shutdown() {
+	request := &client.BulkableRequest{}
+	docsCount := 1000
+	parallelFactor := 10
+	futures := make([]future.Future[bool], docsCount)
+
+	wg := sync.WaitGroup{}
+	wg.Add(parallelFactor + 1) // +1 for separate shutdown goroutine
+	s.mockBulkProcessor.EXPECT().Add(request).MaxTimes(docsCount)
+	s.mockBulkProcessor.EXPECT().Stop().Return(nil).Times(1)
+	s.mockMetricHandler.EXPECT().Timer(metrics.ElasticsearchBulkProcessorWaitAddLatency.GetMetricName()).Return(metrics.NoopTimerMetricFunc).MaxTimes(docsCount)
+	for i := 0; i < parallelFactor; i++ {
+		go func(i int) {
+			for j := 0; j < docsCount/parallelFactor; j++ {
+				futures[i*docsCount/parallelFactor+j] = s.esProcessor.Add(request, fmt.Sprintf("test-key-%d-%d", i, j))
+			}
+			wg.Done()
+		}(i)
+	}
+	go func() {
+		time.Sleep(1 * time.Millisecond) // slight delay so at least a few docs get added
+		s.esProcessor.Stop()
+		wg.Done()
+	}()
+
+	wg.Wait()
+
+	s.Nil(s.esProcessor.mapToAckFuture)
+	s.False(futures[0].Ready())          // first requests should be in bulk
+	s.True(futures[docsCount-1].Ready()) // final requests should be only errors
```

Comment:
The test can't guarantee this I think?

Reply:
It's true it can't be guaranteed since it is a test of concurrency. I think in our test environments we can always expect at least the first doc to be added successfully and the last doc to be added after shutdown. If these two things do not occur, then the test did not actually test a call to Stop in the middle of active calls to Add so the test did not function as expected. Maybe there is some other test case I could use to reduce the chance for flakiness?

Reply:
We need to fix all flaky tests, otherwise it will be very annoying when those tests are failing in buildkite/CICD. 

To me, it's ok that in rare cases tests is not testing the case we have in mind. Because those tests are keep getting exercised and in most cases it's doing what we want and will fail if things go wrong.